File: experiments/runs/2026-09-20_10-25-20_claim-office-java-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking/src/main/java/PolicyWideModifiers.java
import java.math.BigDecimal;

/**
 * The MHPCO's policy-wide premium modifiers. Unlike the item-specific risk
 * surcharges, these are measured against the policy premium as a whole and
 * depend on the customer's history with the MHPCO rather than on any item.
 *
 * This is the one place that says which modifiers are policy-wide: a further
 * policy-wide modifier is added here and nowhere else. So far the MHPCO grants
 * the loyalty discount and charges the initial assessment surcharge.
 */
public final class PolicyWideModifiers {

    private static final int LOYALTY_YEARS = 2;

    private static final int LOYALTY_DISCOUNT_PERCENT = 20;

    private static final int INITIAL_ASSESSMENT_SURCHARGE_PERCENT = 10;

    private static final int FOLLOW_UP_CONTRACT_DISCOUNT_PERCENT = 15;

    private final Customer customer;

    private int contractsIssued;

    public PolicyWideModifiers(Customer customer) {
        this.customer = customer;
    }

    /**
     * The net effect of every policy-wide modifier on a policy premium:
     * positive amounts are charged to the customer, negative amounts are
     * granted to the customer. Each modifier states its own direction, so this
     * sum says only which modifiers the MHPCO applies policy-wide.
     */
    public BigDecimal totalPolicyModifiers(BigDecimal policyPremium) {
        return initialAssessmentSurcharge(policyPremium)
                .add(loyaltyDiscount(policyPremium))
                .add(followUpContractDiscount(policyPremium));
    }

    /**
     * The MHPCO notes that a contract has been issued to this customer. Every
     * later contract is a follow-up contract, which is what the follow-up
     * contract discount is measured against.
     */
    public void recordContractIssued() {
        contractsIssued++;
    }

    /**
     * Customers receive a 15 % discount on each contract after their first.
     */
    private BigDecimal followUpContractDiscount(BigDecimal policyPremium) {
        if (!hasAPreviousContract()) {
            return BigDecimal.ZERO;
        }
        return Percentage.of(policyPremium, FOLLOW_UP_CONTRACT_DISCOUNT_PERCENT).negate();
    }

    private boolean hasAPreviousContract() {
        return contractsIssued > 0;
    }

    /**
     * Long-standing customers -- 2 or more years of business with the MHPCO --
     * receive a 20 % loyalty discount on the policy premium.
     */
    private BigDecimal loyaltyDiscount(BigDecimal policyPremium) {
        if (!isLongStandingCustomer()) {
            return BigDecimal.ZERO;
        }
        return Percentage.of(policyPremium, LOYALTY_DISCOUNT_PERCENT).negate();
    }

    /**
     * The MHPCO counts a customer as long-standing from 2 full years of
     * business onwards; a customer in their second year is still a newcomer.
     */
    private boolean isLongStandingCustomer() {
        return customer.yearsWithMHPCO() >= LOYALTY_YEARS;
    }

    /**
     * Every item in a quote is treated as a first insurance, regardless of the
     * customer's history with the MHPCO, and carries a 10 % initial assessment
     * surcharge on the policy premium.
     */
    private BigDecimal initialAssessmentSurcharge(BigDecimal policyPremium) {
        return Percentage.of(policyPremium, INITIAL_ASSESSMENT_SURCHARGE_PERCENT);
    }

}

File: experiments/runs/2026-09-23_13-26-00_claim-office-example-mapping_exact-ptdd-v1.1-refactor-subagent-cc_opus-5-no-thinking-2/src/policy-wide-modifiers.ts
export interface Customer {
  yearsWithMHPCO: number;
}

const FIRST_INSURANCE_SURCHARGE_RATE = 0.1;
const LOYALTY_DISCOUNT_RATE = 0.2;
const LOYALTY_YEARS = 2;
const FOLLOW_UP_DISCOUNT_RATE = 0.15;

/**
 * The customer's standing with the MHPCO at the moment of a quote: how long the
 * business relationship has lasted, and how many contracts preceded this one.
 */
export interface CustomerStanding {
  customer: Customer;
  previousQuoteCount: number;
}

/**
 * The customer's standing at the moment of this quote, as the MHPCO reads it
 * from the scenario so far: the business relationship and the contracts that
 * preceded this one.
 */
export function standingOf(customer: Customer, previousQuoteCount: number): CustomerStanding {
  return { customer, previousQuoteCount };
}

/** Whether this quote follows an earlier contract of the same customer. */
function isFollowUpContract(standing: CustomerStanding): boolean {
  return standing.previousQuoteCount >= 1;
}

/**
 * Every item in a quote is treated as a first insurance, regardless of customer
 * history, so the office levies the initial assessment on every policy.
 */
function isFirstInsurance(): boolean {
  return true;
}

/** Whether the MHPCO regards the customer as long-standing. */
function isLongStanding(standing: CustomerStanding): boolean {
  return standing.customer.yearsWithMHPCO >= LOYALTY_YEARS;
}

/**
 * A policy-wide clause of the MHPCO's rulebook: the condition under which the
 * office grants or levies it, and the signed rate it charges on the policy base
 * premium — positive for a surcharge, negative for a discount. Condition and
 * rate belong to one clause and change together.
 */
interface PolicyWideClause {
  appliesTo(standing: CustomerStanding): boolean;
  rate: number;
}

/**
 * The policy-wide clauses the MHPCO's rulebook currently contains: the initial
 * assessment surcharge levied on every first insurance, the loyalty discount
 * granted to a long-standing customer, and the follow-up discount granted on
 * each contract after the customer's first.
 */
const POLICY_WIDE_CLAUSES: PolicyWideClause[] = [
  { appliesTo: isFirstInsurance, rate: FIRST_INSURANCE_SURCHARGE_RATE },
  { appliesTo: isLongStanding, rate: -LOYALTY_DISCOUNT_RATE },
  { appliesTo: isFollowUpContract, rate: -FOLLOW_UP_DISCOUNT_RATE },
];

/**
 * The net total of the MHPCO's policy-wide modifiers, each levied on the policy
 * base premium and never on an already-modified amount.
 */
export function policyWideModifiers(policyBase: number, standing: CustomerStanding): number {
  return POLICY_WIDE_CLAUSES.filter((clause) => clause.appliesTo(standing)).reduce(
    (running, clause) => running + policyBase * clause.rate,
    0,
  );
}
